### ISSUE TITLE
vim: ignore `vimContinue` lines?!

### ISSUE THREAD

**Issue by blueyed** [roryokane/detectindent#10]
I've came across a plugin, which (probably accidentally) uses indenting for continued lines that does not fit the rest of the file (https://github.com/hynek/vim-python-pep8-indent/blob/master/indent/python.vim#L36-L41):

    let s:block_rules = {
      \ '^\s*elif\>': ['if', 'elif'],
      \ '^\s*else\>': ['if', 'elif', 'for', 'try', 'except'],
      \ '^\s*except\>': ['try', 'except'],
      \ '^\s*finally\>': ['try', 'except', 'else']
      \ }

It uses two spaces for the continued lines, while the rest of the file is aligned with 4 spaces per level.

I think that a fix for this might be to ignore lines that have the `vimContinue` syntax group.

This would mean to have different set of rules for various filetypes maybe - assuming that there are similar rules for other filetypes.

**Comment by roryokane** [comment]
I’m not quite sure what you mean by “continued lines that do not fit the rest of the file”, or what the `block_rules` section has to do with it. I installed [vim-python-pep8-indent](https://github.com/hynek/vim-python-pep8-indent) and tried loading some code with `if`, `else`, `try`, `except`, and `finally`, but DetectIndent detects the indentation of that code correctly.

By “indenting for continued lines”, do you perhaps mean the way that plugin indents function parameters in long statements that continue to the next line? Here is some code with that type of indentation:

```python
class TestPerson(unittest.TestCase):
    def test_blathering(self):
        self.assertEqual("Hello! It's nice weather we're having, isn't it.",
                         self.person.blather())
```

I do see that DetectIndent detects the indentation of that example incorrectly. The last line starts with 25 spaces, which does not share a common denominator with any other line, so `'shiftwidth'` is incorrectly set to 1. Is that the type of problem you are referring to?

**Comment by blueyed** [comment]
> Is that the type of problem you are referring to?

No.  I was referring to the example code provided above itself, from the plugin and not generated by the plugin.

(But the problem you've found is interesting as well.)

**Comment by roryokane** [comment]
Okay, I understand the problem now.

Both the problem you posted and the problem I posted would be fixed by switching to a compare-lines algorithm as I describe in #11. I just ran both [that file](https://raw.githubusercontent.com/hynek/vim-python-pep8-indent/master/indent/python.vim) and my code example through the JavaScript tool [`detect-indent`](https://github.com/sindresorhus/detect-indent), which uses that algorithm, and it detected the indentation correctly.

There would be some value in special-casing `vimContinue` lines, in addition to switching algorithms. It wouldn’t make a difference in that file you pointed to, but it would make a difference if you edited a file containing just the snippet you posted. However, I don’t think it would be worth implementing such a special case unless someone finds a file in the wild, which one might plausibly edit in Vim, that would be detected incorrectly.

There is the question of whether to quickly fix this issue before implementing #11, and then remove it after the new algorithm makes it unnecessary, so that this problem is avoided in the meantime. While fixing this would be simpler than implementing the new algorithm, I don’t think it’s worth doing, since this style of indentation seems rare. As you say, that formatting was probably an accident. DetectIndent itself uses four-space indents for continued lines, and it chose that style even before that choice would have affected how DetectIndent indented its own file.

Thus, I will close this issue. Progress towards fixing this issue will be tracked by #11. If you find a file in the wild with this type of problem that wouldn’t be detected correctly even with the new algorithm, then post it in a comment, and I will reopen this issue as a reminder to special-case this problem after implementing #11.
